File: sdks/python/src/archon/telemetry.py
"""Cloud telemetry exporter — uploads agent runs to the Archon dashboard.

When ``ARCHON_API_KEY`` is set in the environment (or passed explicitly),
the SDK uploads each completed run to the Archon web dashboard so teams
can review traces, costs, and tier breakdowns without shipping logs.

Design principles:

* **Off by default.** No ``ARCHON_API_KEY`` → no network traffic.
  Local SQLite tracing continues to work unchanged.
* **Never crash the agent.** All network errors are caught and logged.
  A failing dashboard must not break a production agent run.
* **Fire-and-forget.** Uploads happen on a background asyncio task so
  ``Agent.run()`` returns immediately. Callers can ``await`` ``flush()``
  before process exit to drain the queue.
* **Exponential backoff.** Transient 5xx / network errors retry up to
  three times with jitter. 401/403 responses fail fast — a bad key
  won't succeed on retry.

Wire protocol is documented in ``archon-webapp/convex/http.ts`` and the
Next.js proxy at ``archon-webapp/src/app/api/ingest/route.ts``.
"""

from __future__ import annotations

import asyncio
import logging
import os
import random
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

import httpx

if TYPE_CHECKING:
    from archon.types import AgentResult

# ── Configuration constants ───────────────────────────

DEFAULT_BASE_URL = "https://archon.yashbogam.me"
INGEST_PATH = "/api/ingest"
VALIDATE_PATH = "/api/validate"

# Retry tuning — conservative so we never hammer the API on outages.
MAX_RETRIES = 3
INITIAL_BACKOFF_SECONDS = 0.5
BACKOFF_MULTIPLIER = 2.0
MAX_BACKOFF_SECONDS = 8.0
REQUEST_TIMEOUT_SECONDS = 10.0

# SDK identifier sent with every request (helps debug version-specific issues).
USER_AGENT = "archon-python/0.2.0"

# Status codes that should not be retried (auth / bad request).
NO_RETRY_STATUS = frozenset({400, 401, 403, 404, 422})

logger = logging.getLogger("archon.telemetry")


# ── Validation result ─────────────────────────────────

@dataclass(frozen=True)
class ValidationResult:
    """Outcome of a ``GET /api/validate`` call."""

    valid: bool
    scopes: tuple[str, ...] = ()
    expires_at: int | None = None
    error: str | None = None


# ── Public helpers ────────────────────────────────────

def resolve_api_key(explicit: str | None = None) -> str | None:
    """Pick the API key from the explicit arg or the ``ARCHON_API_KEY`` env var.

    Returns ``None`` if neither is set — the telemetry exporter treats
    that as "disabled" rather than an error, matching the local-first
    design of the rest of the SDK.
    """
    key = explicit or os.environ.get("ARCHON_API_KEY")
    if not key:
        return None
    key = key.strip()
    return key or None


def resolve_base_url(explicit: str | None = None) -> str:
    """Pick the dashboard base URL, preferring explicit > env > default."""
    url = explicit or os.environ.get("ARCHON_BASE_URL") or DEFAULT_BASE_URL
    return url.rstrip("/")


# ── Telemetry client ──────────────────────────────────

class TelemetryClient:
    """Async client that uploads ``AgentResult`` objects to the dashboard.

    Instantiate once per Agent (or share across agents in the same process).
    The client lazily starts an ``httpx.AsyncClient`` and a background
    task queue on first use, so constructing a ``TelemetryClient`` without
    ever calling ``upload_run`` has zero runtime cost.

    Args:
        api_key: The ``arc_...`` key. If ``None``, the client is disabled
            and all operations become no-ops. This is the default behavior
            when ``ARCHON_API_KEY`` is not set.
        base_url: Dashboard base URL. Defaults to the env var or the
            production hosted dashboard.
    """

    def __init__(
        self,
        api_key: str | None = None,
        *,
        base_url: str | None = None,
    ) -> None:
        self._api_key = resolve_api_key(api_key)
        self._base_url = resolve_base_url(base_url)
        self._client: httpx.AsyncClient | None = None
        self._pending: set[asyncio.Task[None]] = set()

    @property
    def enabled(self) -> bool:
        """True when an API key is available and uploads will be attempted."""
        return self._api_key is not None

    # ── Validation ─────────────────────────────────────

    async def validate(self) -> ValidationResult:
        """Call ``GET /api/validate`` to verify the API key.

        Safe to call without a key — returns ``ValidationResult(valid=False)``
        with an ``error`` message in that case. Never raises.
        """
        if not self._api_key:
            return ValidationResult(valid=False, error="ARCHON_API_KEY is not set")

        client = self._ensure_client()
        try:
            response = await client.get(
                f"{self._base_url}{VALIDATE_PATH}",
                headers=self._auth_headers(),
            )
        except httpx.HTTPError as exc:
            return ValidationResult(valid=False, error=f"Network error: {exc}")

        try:
            data = response.json()
        except ValueError:
            return ValidationResult(
                valid=False, error=f"Invalid response ({response.status_code})"
            )

        if response.status_code == 200 and data.get("valid"):
            return ValidationResult(
                valid=True,
                scopes=tuple(data.get("scopes", [])),
                expires_at=data.get("expiresAt"),
            )
        return ValidationResult(valid=False, error=data.get("error", "Unknown error"))

    # ── Upload API ─────────────────────────────────────

    def upload_run(
        self,
        result: AgentResult,
        *,
        agent_name: str,
    ) -> None:
        """Schedule an upload for ``result`` without blocking the caller.

        The actual HTTP request happens on a background asyncio task.
        If no event loop is running (sync caller), falls back to a
        one-shot ``asyncio.run`` in a worker thread via
        ``asyncio.get_event_loop().create_task`` semantics.

        This is fire-and-forget by design. Call ``await flush()`` if
        you need to guarantee delivery before process exit.
        """
        if not self.enabled:
            return

        payload = _payload_from_result(result, agent_name=agent_name)

        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            # No running loop — run synchronously. This path is rare because
            # ``Agent.run`` is already async, but covers sync test harnesses.
            asyncio.run(self._send_with_retry(payload))
            return

        task = loop.create_task(self._send_with_retry(payload))
        self._pending.add(task)
        task.add_done_callback(self._pending.discard)

    async def flush(self) -> None:
        """Wait for all in-flight uploads to complete."""
        if not self._pending:
            return
        await asyncio.gather(*self._pending, return_exceptions=True)

    async def aclose(self) -> None:
        """Close the underlying HTTP client after flushing pending uploads."""
        await self.flush()
        if self._client is not None:
            await self._client.aclose()
            self._client = None

    # ── Private helpers ────────────────────────────────

    def _auth_headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self._api_key}",
            "Content-Type": "application/json",
            "User-Agent": USER_AGENT,
        }

    def _ensure_client(self) -> httpx.AsyncClient:
        if self._client is None:
            self._client = httpx.AsyncClient(timeout=REQUEST_TIMEOUT_SECONDS)
        return self._client

    async def _send_with_retry(self, payload: dict[str, Any]) -> None:
        """POST ``payload`` to ``/api/ingest`` with bounded retries."""
        client = self._ensure_client()
        url = f"{self._base_url}{INGEST_PATH}"
        backoff = INITIAL_BACKOFF_SECONDS

        for attempt in range(1, MAX_RETRIES + 1):
            try:
                response = await client.post(
                    url, json=payload, headers=self._auth_headers(),
                )
            except httpx.HTTPError as exc:
                logger.debug("archon telemetry network error (attempt %d): %s", attempt, exc)
                if attempt >= MAX_RETRIES:
                    logger.warning(
                        "archon telemetry upload failed after %d attempts: %s", attempt, exc,
                    )
                    return
                await _sleep_with_jitter(backoff)
                backoff = min(backoff * BACKOFF_MULTIPLIER, MAX_BACKOFF_SECONDS)
                continue

            if response.status_code == 200:
                return
            if response.status_code in NO_RETRY_STATUS:
                logger.warning(
                    "archon telemetry rejected: %d %s",
                    response.status_code,
                    _safe_error_body(response),
                )
                return
            # 5xx / other — retry
            logger.debug(
                "archon telemetry retryable status %d (attempt %d)",
                response.status_code, attempt,
            )
            if attempt >= MAX_RETRIES:
                logger.warning(
                    "archon telemetry upload failed after %d attempts: %d",
                    attempt, response.status_code,
                )
                return
            await _sleep_with_jitter(backoff)
            backoff = min(backoff * BACKOFF_MULTIPLIER, MAX_BACKOFF_SECONDS)


# ── Payload construction ──────────────────────────────

def _payload_from_result(result: AgentResult, *, agent_name: str) -> dict[str, Any]:
    """Convert an ``AgentResult`` into the JSON shape the webapp expects.

    Contract defined in ``archon-webapp/convex/http.ts``. Keep in sync.
    """
    steps_payload = []
    for index, step in enumerate(result.steps, start=1):
        steps_payload.append({
            "step": index,
            "gate": "cache" if step.cached else "execute",
            "action": step.tool_call or "llm_call",
            "result": "cached" if step.cached else "ok",
            "duration": step.latency_ms,
            "cost": step.cost_usd,
            "model": step.model,
            "tokens": step.input_tokens + step.output_tokens,
        })

    first_model = result.steps[0].model if result.steps else "unknown"
    first_tier = result.steps[0].tier.value if result.steps else "standard"
    total_tokens = result.total_input_tokens + result.total_output_tokens

    return {
        "runId": result.run_id,
        "agent": agent_name,
        "model": first_model,
        "tier": first_tier,
        "steps": steps_payload,
        "totalCost": result.total_cost_usd,
        "totalTokens": total_tokens,
        "inputTokens": result.total_input_tokens,
        "outputTokens": result.total_output_tokens,
        "latency": result.total_latency_ms,
        "status": "success" if result.output and not result.output.startswith("[") else "error",
        "startedAt": result.started_at.isoformat(),
    }


# ── Misc utilities ────────────────────────────────────

async def _sleep_with_jitter(seconds: float) -> None:
    """Sleep with ±20% jitter to avoid thundering-herd retries."""
    # ``random`` is fine here — this is jitter, not a cryptographic use.
    jitter = seconds * 0.2 * (2.0 * random.random() - 1.0)  # noqa: S311
    await asyncio.sleep(max(0.0, seconds + jitter))


def _safe_error_body(response: httpx.Response) -> str:
    """Extract a short error string from a response without crashing."""
    try:
        data = response.json()
        if isinstance(data, dict) and "error" in data:
            return str(data["error"])[:200]
    except ValueError:
        pass
    return response.text[:200] if response.text else ""

File: sdks/typescript/src/telemetry.ts
/**
 * Cloud telemetry exporter — uploads agent runs to the Archon dashboard.
 *
 * When `ARCHON_API_KEY` is set (or `apiKey` is passed explicitly), the
 * SDK uploads each completed run to the Archon web dashboard. Local
 * execution and local tracing continue to work unchanged when the env
 * var is missing — this module is a pure additive layer.
 *
 * Design principles:
 *   - **Off by default.** No `ARCHON_API_KEY` → no network traffic.
 *   - **Never crash the agent.** All errors are caught and logged.
 *   - **Fire-and-forget.** Uploads run on a background Promise.
 *     Call `await flush()` before process exit to drain the queue.
 *   - **Exponential backoff.** Transient 5xx / network errors retry
 *     up to three times with jitter. 4xx fails fast.
 *
 * Wire protocol matches `archon-webapp/convex/http.ts` and the Next.js
 * proxy at `archon-webapp/src/app/api/ingest/route.ts`.
 */

import type { AgentResult } from "./types.js";

// ── Configuration constants ───────────────────────────

const DEFAULT_BASE_URL = "https://archon.yashbogam.me";
const INGEST_PATH = "/api/ingest";
const VALIDATE_PATH = "/api/validate";

/** SDK identifier sent with every request. */
const USER_AGENT = "archon-typescript/0.1.0";

/** Retry tuning — conservative so we never hammer the API on outages. */
const MAX_RETRIES = 3;
const INITIAL_BACKOFF_MS = 500;
const BACKOFF_MULTIPLIER = 2;
const MAX_BACKOFF_MS = 8000;
const REQUEST_TIMEOUT_MS = 10_000;

/** Status codes that should not be retried (auth / bad request). */
const NO_RETRY_STATUS = new Set([400, 401, 403, 404, 422]);

// ── Public types ──────────────────────────────────────

export interface TelemetryClientConfig {
  /**
   * The `arc_...` key. If omitted, falls back to `process.env.ARCHON_API_KEY`.
   * If still absent, the client stays disabled and all operations are no-ops.
   */
  apiKey?: string;
  /** Dashboard base URL. Falls back to `ARCHON_BASE_URL`, then the default. */
  baseUrl?: string;
  /** Optional logger. Defaults to `console`. */
  logger?: Pick<Console, "warn" | "debug">;
}

export interface ValidationResult {
  valid: boolean;
  scopes?: readonly string[];
  expiresAt?: number | null;
  error?: string;
}

// ── Public helpers ────────────────────────────────────

/** Pick the API key from the explicit arg or the `ARCHON_API_KEY` env var. */
export function resolveApiKey(explicit?: string): string | undefined {
  const raw = explicit ?? process.env.ARCHON_API_KEY;
  const trimmed = raw?.trim();
  return trimmed ? trimmed : undefined;
}

/** Pick the dashboard base URL: explicit > env > default. */
export function resolveBaseUrl(explicit?: string): string {
  const raw = explicit ?? process.env.ARCHON_BASE_URL ?? DEFAULT_BASE_URL;
  return raw.replace(/\/+$/, "");
}

// ── Telemetry client ──────────────────────────────────

export class TelemetryClient {
  readonly #apiKey: string | undefined;
  readonly #baseUrl: string;
  readonly #logger: Pick<Console, "warn" | "debug">;
  readonly #pending = new Set<Promise<void>>();

  constructor(config: TelemetryClientConfig = {}) {
    this.#apiKey = resolveApiKey(config.apiKey);
    this.#baseUrl = resolveBaseUrl(config.baseUrl);
    this.#logger = config.logger ?? console;
  }

  /** True when an API key is available and uploads will be attempted. */
  get enabled(): boolean {
    return this.#apiKey !== undefined;
  }

  /**
   * Verify the API key via `GET /api/validate`. Never throws — returns
   * a `ValidationResult` with an `error` message on failure.
   */
  async validate(): Promise<ValidationResult> {
    if (!this.#apiKey) {
      return { valid: false, error: "ARCHON_API_KEY is not set" };
    }

    let response: Response;
    try {
      response = await this.#fetchWithTimeout(
        `${this.#baseUrl}${VALIDATE_PATH}`,
        { method: "GET", headers: this.#authHeaders() },
      );
    } catch (err) {
      return { valid: false, error: `Network error: ${(err as Error).message}` };
    }

    let data: unknown;
    try {
      data = await response.json();
    } catch {
      return { valid: false, error: `Invalid response (${response.status})` };
    }

    const d = data as {
      valid?: boolean;
      scopes?: string[];
      expiresAt?: number | null;
      error?: string;
    };
    if (response.status === 200 && d.valid) {
      return { valid: true, scopes: d.scopes ?? [], expiresAt: d.expiresAt ?? null };
    }
    return { valid: false, error: d.error ?? "Unknown error" };
  }

  /**
   * Schedule an upload for `result` without blocking the caller.
   * The HTTP request runs on a background Promise. Call `flush()` to
   * wait for in-flight uploads to drain before exit.
   */
  uploadRun(result: AgentResult, agentName: string): void {
    if (!this.enabled) return;
    const payload = payloadFromResult(result, agentName);
    const task = this.#sendWithRetry(payload).catch((err) => {
      // This should never happen — sendWithRetry already catches
      // everything — but belt-and-suspenders: the agent must not crash
      // because telemetry failed.
      this.#logger.warn("archon telemetry: uncaught upload error", err);
    });
    this.#pending.add(task);
    void task.finally(() => this.#pending.delete(task));
  }

  /** Wait for all in-flight uploads to complete. */
  async flush(): Promise<void> {
    if (this.#pending.size === 0) return;
    await Promise.allSettled([...this.#pending]);
  }

  // ── Private helpers ─────────────────────────────────

  #authHeaders(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.#apiKey}`,
      "Content-Type": "application/json",
      "User-Agent": USER_AGENT,
    };
  }

  async #fetchWithTimeout(url: string, init: RequestInit): Promise<Response> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
    try {
      return await fetch(url, { ...init, signal: controller.signal });
    } finally {
      clearTimeout(timeout);
    }
  }

  async #sendWithRetry(payload: IngestPayload): Promise<void> {
    const url = `${this.#baseUrl}${INGEST_PATH}`;
    let backoff = INITIAL_BACKOFF_MS;

    for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
      let response: Response;
      try {
        response = await this.#fetchWithTimeout(url, {
          method: "POST",
          headers: this.#authHeaders(),
          body: JSON.stringify(payload),
        });
      } catch (err) {
        this.#logger.debug(
          `archon telemetry network error (attempt ${attempt}): ${(err as Error).message}`,
        );
        if (attempt >= MAX_RETRIES) {
          this.#logger.warn(
            `archon telemetry upload failed after ${attempt} attempts: ${(err as Error).message}`,
          );
          return;
        }
        await sleepWithJitter(backoff);
        backoff = Math.min(backoff * BACKOFF_MULTIPLIER, MAX_BACKOFF_MS);
        continue;
      }

      if (response.status === 200) return;
      if (NO_RETRY_STATUS.has(response.status)) {
        this.#logger.warn(
          `archon telemetry rejected: ${response.status} ${await safeErrorBody(response)}`,
        );
        return;
      }
      // 5xx or transient — retry.
      this.#logger.debug(
        `archon telemetry retryable status ${response.status} (attempt ${attempt})`,
      );
      if (attempt >= MAX_RETRIES) {
        this.#logger.warn(
          `archon telemetry upload failed after ${attempt} attempts: ${response.status}`,
        );
        return;
      }
      await sleepWithJitter(backoff);
      backoff = Math.min(backoff * BACKOFF_MULTIPLIER, MAX_BACKOFF_MS);
    }
  }
}

// ── Payload construction ──────────────────────────────

interface IngestStep {
  step: number;
  gate: string;
  action: string;
  result: string;
  duration: number;
  cost: number;
  model?: string;
  tokens?: number;
}

interface IngestPayload {
  runId: string;
  agent: string;
  model: string;
  tier: string;
  steps: IngestStep[];
  totalCost: number;
  totalTokens: number;
  inputTokens: number;
  outputTokens: number;
  latency: number;
  status: string;
  startedAt: string;
}

function payloadFromResult(result: AgentResult, agentName: string): IngestPayload {
  const steps: IngestStep[] = result.steps.map((step, index) => ({
    step: index + 1,
    gate: step.cached ? "cache" : "execute",
    action: step.toolCall ?? "llm_call",
    result: step.cached ? "cached" : "ok",
    duration: step.latencyMs,
    cost: step.costUsd,
    model: step.model,
    tokens: step.inputTokens + step.outputTokens,
  }));

  const firstStep = result.steps[0];
  const firstModel = firstStep?.model ?? "unknown";
  const firstTier = firstStep?.tier ?? "standard";
  const totalTokens = result.totalInputTokens + result.totalOutputTokens;

  // An output string starting with "[" is how the agent reports errors
  // (e.g. "[Budget exceeded ...]"). Treat empty output the same way so
  // the dashboard can flag failed runs.
  const status = result.output && !result.output.startsWith("[") ? "success" : "error";

  return {
    runId: result.runId,
    agent: agentName,
    model: firstModel,
    tier: firstTier,
    steps,
    totalCost: result.totalCostUsd,
    totalTokens,
    inputTokens: result.totalInputTokens,
    outputTokens: result.totalOutputTokens,
    latency: result.totalLatencyMs,
    status,
    startedAt: result.startedAt.toISOString(),
  };
}

// ── Misc utilities ────────────────────────────────────

async function sleepWithJitter(ms: number): Promise<void> {
  // ±20% jitter to avoid thundering-herd retries.
  const jitter = ms * 0.2 * (2 * Math.random() - 1);
  const wait = Math.max(0, ms + jitter);
  await new Promise((resolve) => setTimeout(resolve, wait));
}

async function safeErrorBody(response: Response): Promise<string> {
  try {
    const data = (await response.json()) as { error?: string };
    if (data && typeof data.error === "string") return data.error.slice(0, 200);
  } catch {
    /* fall through */
  }
  try {
    const text = await response.text();
    return text.slice(0, 200);
  } catch {
    return "";
  }
}
